staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -40,6 +43,10 @@ type cachedTokenAuthenticator struct {
 	failureTTL time.Duration
 
 	cache cache
+
+	// randomCachePrefix is a per authenticator random value that is used to prevent precomputation attacks
+	// It also mitigates hash map DOS attacks (the inputs are supplied by untrusted users)
```

Comment:
DOS attacks or collision attacks?

Reply:
Same thing?  i.e. creating keys that purposefully collide limits the usefulness of the hash map == DOS of hash map?

Reply:
Collision attacks on hash maps is a dos since it degrades hash map performance to worst case.

Reply:
Collision also means you get the attributes of whatever was in the cache, which seems worse than just a dos to me?

Reply:
It's worse in outcome but not practical given that sha256 is still considered collision resistant, and tokens should already embed sufficient entropy (e.g. service account tokens have two uuids with 122-bits of random each). This primarily protects us from collisions in the hash function used by map[] which is not collision resistant by concealing the hashing result (which becomes the key to the map[] hash function) inside the apiserver.

Reply:
Just so I'm clear, you're suggesting folks could do a birthday attack on the key hashes to find ones that collide via go map's internal hash function? And then send enough of them to make a linear search of the number of keys a significant factor in the overall request response time?

That doesn't seem like a plausible attack here (not enough tokens and/or linear is not terrible), but I like that level of paranoia.

Reply:
It's definitely paranoia :) but it's cropped up in research and CVEs since early 2000s.

https://www.usenix.org/legacy/events/sec03/tech/full_papers/crosby/crosby_html/index.html
https://github.com/google/highwayhash
https://arxiv.org/pdf/1612.06257.pdf
https://en.wikipedia.org/wiki/SipHash